### ISSUE TITLE
Clang does not detect narrowing conversion with const ref

### ISSUE THREAD

**Issue by estan** [llvm/llvm-project#63151]
Clang compiles

```cpp
struct A { A(const unsigned &x) {} };

int main(void) {
    A a { -1 };
    return 0;
}
```

without emitting any diagnostic, but GCC emits

```console
test.cpp: In function ‘int main()’:
test.cpp:4:14: error: narrowing conversion of ‘-1’ from ‘int’ to ‘unsigned int’ [-Wnarrowing]
    4 |     A a { -1 };
      |
```

I think it is correct to treat the program as ill formed in this case, but I'm not sure.

**Comment by AaronBallman** [comment]
> I think it is correct to treat the program as ill formed in this case, but I'm not sure.

Yes, this is a narrowing conversion per http://eel.is/c++draft/dcl.init.list#7.4 The fact that it's a reference doesn't change things.

**Comment by estan** [comment]
@AaronBallman Thanks for confirming. I mentioned the const reffedness since it emits diagnostic for

```cpp
struct A { A(unsigned x) {} };

int main(void) {
    A a { -1 };
    return 0;
}
```

**Comment by AaronBallman** [comment]
Yeah, it surprised me when I noticed the same thing. I only mentioned it because I had to go make sure that a temp reference binding didn't change the answer.
